File: clj/src/h3m_lwp_clj/assets.clj
(ns h3m-lwp-clj.assets
  (:import [com.badlogic.gdx.graphics Texture]
           [com.badlogic.gdx.graphics.g2d TextureRegion TextureAtlas TextureAtlas$AtlasRegion]
           [com.badlogic.gdx.assets AssetManager]
           [com.badlogic.gdx.assets.loaders TextureAtlasLoader$TextureAtlasParameter]
           [com.badlogic.gdx.utils Array]))


(def ^String objects-atlas "sprites/all.atlas")


(defonce objects-info (atom {}))


(def ^AssetManager manager
  (doto (new AssetManager)
    (.load objects-atlas TextureAtlas (new TextureAtlasLoader$TextureAtlasParameter true))))


(defn init []
  (reset!
   objects-info
   (try
     ; TODO AssetLoader?
     (read-string (slurp "sprites/all.edn"))
     (catch Exception _
       (println "Failed to read all.edn")
       {})))
  (.finishLoading manager)
  (Texture/setAssetManager manager))


(defn get-sprite-info [def-name]
  (get @objects-info def-name))


(defn get-object-frame [def-name offset]
  ^TextureRegion
  (let [^TextureAtlas atlas (.get manager objects-atlas)
        ^TextureAtlas$AtlasRegion frame (.findRegion atlas def-name offset)]
    (when (nil? frame)
      (throw (new Exception (format "def %s with offset %d not found in atlas" def-name offset))))
    frame))


(defn get-terrain-sprite [def-name index]
  ^Array
  (let [^TextureAtlas atlas (.get manager objects-atlas)
        sprite-info (get-sprite-info def-name)
        {frames-order :order} sprite-info
        offset (nth frames-order index)
        region-name (format "%s/%d" def-name offset)
        ^Array frames (.findRegions atlas region-name)]
    (when (.isEmpty frames)
      (throw (new Exception (format "def %s with index %d not found in atlas" def-name index))))
    frames))

File: core/src/com/homm3/livewallpaper/core/Assets.kt
package com.homm3.livewallpaper.core

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.assets.AssetManager
import com.badlogic.gdx.assets.loaders.TextureAtlasLoader
import com.badlogic.gdx.assets.loaders.resolvers.LocalFileHandleResolver
import com.badlogic.gdx.graphics.Pixmap
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.scenes.scene2d.ui.Skin
import com.badlogic.gdx.utils.Array
import ktx.assets.load
import ktx.collections.gdxArrayOf
import java.util.*

class Assets {
    companion object {
        const val atlasFolder = "sprites"
        const val atlasName = "images"
        const val atlasPath = "$atlasFolder/$atlasName.atlas"
        const val skinPath = "skin/uiskin.json"
    }

    private val emptyPixmap = TextureAtlas.AtlasRegion(
        TextureRegion(
            Texture(
                Pixmap(0, 0, Pixmap.Format.RGBA8888)
            )
        )
    )
    val manager = AssetManager(LocalFileHandleResolver())

    init {
        Texture.setAssetManager(manager)
    }

    private fun getAtlasRegions(defName: String): Array<TextureAtlas.AtlasRegion> {
        val regions = manager
            .get<TextureAtlas>(atlasPath)
            .findRegions(defName)

        if (regions.isEmpty) {
            println("Can't find def $defName")
            return gdxArrayOf(emptyPixmap)
        }

        return regions
    }

    fun getObjectFrames(defName: String): Array<TextureAtlas.AtlasRegion> {
        return getAtlasRegions(defName.toLowerCase(Locale.ROOT).removeSuffix(".def"))
    }

    fun getTerrainFrames(defName: String, index: Int): Array<TextureAtlas.AtlasRegion> {
        return getAtlasRegions("$defName/$index")
    }

    fun loadSkin(): Skin {
        return Skin(Gdx.files.internal(skinPath))
    }

    fun isWallpaperAssetsLoaded(): Boolean {
        return manager.isLoaded(atlasPath)
    }

    private fun canLoadWallpapersAssets(): Boolean {
        val isAssetsReady = Gdx.app
            .getPreferences(Engine.PREFERENCES_NAME)
            .getBoolean(Engine.IS_ASSETS_READY_KEY)
        val filesExists = Gdx.files.local(atlasPath).exists()

        return isAssetsReady && filesExists && !isWallpaperAssetsLoaded()
    }

    fun tryLoadWallpaperAssets() {
        if (!manager.contains(atlasPath) && canLoadWallpapersAssets()) {
            manager.load<TextureAtlas>(
                atlasPath,
                TextureAtlasLoader.TextureAtlasParameter(true)
            )
        }
    }
}